arangod/Transaction/Methods.cpp

```diff
@@ -1751,162 +2028,128 @@ Future<OperationResult> transaction::Methods::removeLocal(
-            options);
-      }
+  std::shared_ptr<std::vector<ServerID> const> followers;
+  // this call will populate replicationType and followers
+  Result res = determineReplicationTypeAndFollowers(
+      *collection, "remove", value, options, replicationType, followers);
 
-      // we are a valid follower. we do not need to send a proper result with
-      // _key, _id, _rev back to the leader, because it will ignore all these
-      // data anyway. it is sufficient to send headers and the proper error
-      // codes back.
-      options.silent = true;
-    }
-  }  // isDBServer - early block
+  if (res.fail()) {
+    return futures::makeFuture(OperationResult(std::move(res), options));
+  }
 
-  TRI_ASSERT((replicationType == ReplicationType::LEADER) ==
-             (followers != nullptr));
-  TRI_ASSERT(!options.silent || replicationType != ReplicationType::LEADER ||
-             followers->empty());
+  bool excludeAllFromReplication =
+      (replicationType != ReplicationType::LEADER ||
+       (followers->empty() &&
+        collection->replicationVersion() != replication::Version::TWO));
 
+  // total result that is going to be returned to the caller, append-only
   VPackBuilder resultBuilder;
-  ManagedDocumentResult previous;
+  // all document data that are going to be replicated, append-only
+  transaction::BuilderLeaser replicationData(this);
+  // builder for a single, old version of document (will be recycled for each
+  // document)
+  transaction::BuilderLeaser previousDocumentBuilder(this);
+  // temporary builder for building keys
+  transaction::BuilderLeaser keyBuilder(this);
+
+  auto workForOneDocument = [&](VPackSlice value, bool isArray) -> Result {
+    Result res;
 
-  auto workForOneDocument = [&](VPackSlice value, bool isBabies) -> Result {
-    transaction::BuilderLeaser builder(this);
     std::string_view key;
     if (value.isString()) {
       key = value.stringView();
+      // strip everything before a / (likely an _id value)
       size_t pos = key.find('/');
       if (pos != std::string::npos) {
         key = key.substr(pos + 1);
-        builder->add(
-            VPackValuePair(key.data(), key.length(), VPackValueType::String));
-        value = builder->slice();
+        keyBuilder->clear();
+        keyBuilder->add(VPackValue(key));
+        value = keyBuilder->slice();
       }
     } else if (value.isObject()) {
       VPackSlice keySlice = value.get(StaticStrings::KeyString);
       if (!keySlice.isString()) {
-        return Result(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
+        return res.reset(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
       }
       key = keySlice.stringView();
-    } else {
-      return Result(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
     }
 
-    // Primary keys must not be empty
+    // primary key must not be empty
     if (key.empty()) {
-      return Result(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
+      return res.reset(TRI_ERROR_ARANGO_DOCUMENT_HANDLE_BAD);
```

Comment:
Can you explain why you suggest to write 
`return ....reset(...)` instead of `return {...};`

I think it can be better only if we already have an error and replace it with another
But know we don't make optimization in this case, and this case is very rare